### PR TITLE
EVM: Precompile Debug Logger Improvements

### PR DESCRIPTION
**Pull request by holgerd77**
This PR adds some extended logging to the EVM precompiles, WIP.

### CODE REVIEW

**Comment by holgerd77** [comment]
Works! 🤩

This is how things look like (run with `DEBUG=ethjs,evm:precompiles npm run client:start` e.g. if used for client).

grafik

Open for review. 🙂

File: packages/evm/src/precompiles/07-ecmul.ts [hunk 1]
```diff
@@ -8,14 +10,28 @@ const bn128 = require('rustbn.js')
...
     return OOGResult(opts.gasLimit)
   }
 
   const returnData = bn128.mul(inputData)
   // check ecmul success or failure by comparing the output length
   if (returnData.length !== 64) {
+    if (opts._debug) {
+      opts._debug(`ECMUL (0x07) failed: OOG`)
+    }
+    // TODO: should this really return OOG?
```
Comment (jochem-brouwer) [inline packages/evm/src/precompiles/07-ecmul.ts:34]: TODO

File: packages/evm/src/precompiles/08-ecpairing.ts [hunk 2]
```diff
@@ -12,15 +14,29 @@ export function precompile08(opts: PrecompileInput): ExecResult {
...
   }
 
   const returnData = bn128.pairing(inputData)
 
   // check ecpairing success or failure by comparing the output length
   if (returnData.length !== 32) {
+    if (opts._debug) {
+      opts._debug(`ECPAIRING (0x08) failed: OOG`)
+    }
+    // TODO: should this really return OOG?
```
Comment (jochem-brouwer) [inline packages/evm/src/precompiles/08-ecpairing.ts:39]: TODO

File: packages/evm/src/precompiles/0a-bls12-g1add.ts [hunk 3]
```diff
@@ -13,6 +15,13 @@ export async function precompile0a(opts: PrecompileInput): Promise<ExecResult> {
...
+  if (opts._debug) {
+    opts._debug(
+      `Run BLS12G1ADD (0x0a) precompile data=${short(opts.data)} length=${
+        opts.data.length
+      } gasLimit=${opts.gasLimit} gasUsed=${gasUsed}`
+    )
+  }
 
   if (opts.gasLimit < gasUsed) {
     return OOGResult(opts.gasLimit)
```
Comment (jochem-brouwer) [inline packages/evm/src/precompiles/0a-bls12-g1add.ts:27]: Missing OOG debug?

**Review by jochem-brouwer — COMMENTED** [review]
These logs are great, but:

- There are 2 TODOs left
- In BLS precompiles only input is recorded
- I would suggest also adding a debug log for the output buffer / return value of the precompile, in case it succeeds

File: packages/evm/src/precompiles/07-ecmul.ts (hunk 1 above)
Comment (holgerd77) [reply packages/evm/src/precompiles/07-ecmul.ts:34]: I only added these as some other scoped tasks (behavior was there before), no intention to change along this PR.

**Comment by holgerd77** [comment]
> In BLS precompiles only input is recorded

Yeah, I was lazy because BLS is not that much used yet. But can (will) add. 😋

> I would suggest also adding a debug log for the output buffer / return value of the precompile, in case it succeeds

Yes, we can do, makes sense. Quick note: there sometimes are results (e.g. keccak hash) directly computed and returned in the result statement. I would then need to put this a line before and in-between assign this to an additional variable (`const res = ...`). I was a bit sensitive on this due to this not-copy-over topic. But I would agree that for such short buffers this should not make any noticeable performance differences.

**Comment by holgerd77** [comment]
Hmm. Client tests pass locally for me. Will do the changes later the day and then push again and see what happens, maybe that was a one time thing.

File: packages/evm/src/precompiles/07-ecmul.ts (hunk 1 above)
Comment (jochem-brouwer) [reply packages/evm/src/precompiles/07-ecmul.ts:34]: Quick note: I cannot imagine that this is wrong, this should be covered by the test suite.

**Comment by jochem-brouwer** [comment]
This is the error on CI:

```
[03-07|13:36:05] INFO Miner stopped.
/home/runner/work/ethereumjs-monorepo/ethereumjs-monorepo/node_modules/abstract-level/abstract-level.js:806
    db.nextTick(callback, new ModuleError('Database is not open', {
                          ^
ModuleError: Database is not open
    at maybeError (/home/runner/work/ethereumjs-monorepo/ethereumjs-monorepo/node_modules/abstract-level/abstract-level.js:806:27)
    at MemoryLevel.get (/home/runner/work/ethereumjs-monorepo/ethereumjs-monorepo/node_modules/abstract-level/abstract-level.js:282:9)
    at DBManager.get (/home/runner/work/ethereumjs-monorepo/ethereumjs-monorepo/packages/blockchain/src/db/manager.ts:219:32)
    at DBManager.getBody (/home/runner/work/ethereumjs-monorepo/ethereumjs-monorepo/packages/blockchain/src/db/manager.ts:149:29)
    at DBManager.getBlock (/home/runner/work/ethereumjs-monorepo/ethereumjs-monorepo/packages/blockchain/src/db/manager.ts:111:25)
    at async Blockchain.getBlock (/home/runner/work/ethereumjs-monorepo/ethereumjs-monorepo/packages/blockchain/src/blockchain.ts:744:14)
    at async Blockchain.runWithLock (/home/runner/work/ethereumjs-monorepo/ethereumjs-monorepo/packages/blockchain/src/blockchain.ts:295:21)
    at async Chain.update (/home/runner/work/ethereumjs-monorepo/ethereumjs-monorepo/packages/client/lib/blockchain/chain.ts:234:21)
    at async Chain.putBlocks (/home/runner/work/ethereumjs-monorepo/ethereumjs-monorepo/packages/client/lib/blockchain/chain.ts:347:7)
    at async FullSynchronizer.handleNewBlock (/home/runner/work/ethereumjs-monorepo/ethereumjs-monorepo/packages/client/lib/sync/fullsync.ts:326:7) {
  code: 'LEVEL_DATABASE_NOT_OPEN'
}
```

Looks like a race condition to me (blockchain tries to close already closed DB?)

**Comment by holgerd77** [comment]
Ok, have added, we'll see if the tests will pass now. 🙂

**Comment by holgerd77** [comment]
This should be ok now, can this get a review?

**Review by jochem-brouwer — APPROVED** [review]

**Pull request merged**